readme.markdown
# EasyGravatar
__EasyGravatar__ is a simple PHP wrapper class for Gravatar that allows you to get user Avatar or Profile eaisly.


## Requirements

 - __PHP 5.3+__
 - SPL
 - `allow_url_fopen` enabled
 

## Installation
Depends what you want to do you have to download other file (there are two).

  - *src/Image.php* if you want to get user's avatar 
  - *src/Profile.php* if you want to get user's profile data.


If you know which file you want, put it into your `include_path` or `include` path.


## Avatar

### Basic Usage

Simplest way to get user's avatar requires only to call `EasyGravatar\Image::getImage()`.  
Here's some example code:

    # Include Image.php file
     include 'path/to/EasyGravatar/Image.php';
 
    # Creating EasyGravatar\Image object with specifing user's email
     $img = new EasyGravatar\Image('youremailhere@example.com');

    # Get plain image URL
     echo $img->getImage();

Code above will print URL to your avatar.

 > __TIP:__
 > *If you have only hashed user e-mail or want to hash it by your own, while creating object you can pass hashed mail instead of non-hashed, it will work.*


#### Changing result type
EasyGravatar allows you to get user's image and return in in few formats:

  * _EasyGravatar\\Image_::__Plain__ - Returns plain URL to avatar. [default]
  * _EasyGravatar\\Image_::__HTML__ - Returns HTML `<img>` tag with specified `src` attribute as Avatar URL and `alt` as `$alt` which is 'Avatar' by default.
  * _EasyGravatar\\Image_::__BB__ - Returns BB `[img]` tag with specified URL.
  * _EasyGravatar\\Image_::__Markdown__ - Returns image in Markdown syntax. _[Since 0.1.2]_


Example:

	$img->getImage( EasyGravatar\Image::HTML );

Will return:

	<img src="path" alt="$alt value" />


## Request parameters
 You can also get user's avatar with specified size, rating and others.

### Size
 Default size specified by EasyGravatar is `60`, so if you want to get image in this size you can avoid specifing size, but if you want to use other size you can specify avatar size using `EasyGravatar\Image::setSize($size);` method. Some simple example:
 
     $img->setSize(50);
     
But there are also defined simple size constants:

 - _EasyGravatar\\Image_::__SizeSmall__ - 30x30px
 - _EasyGravatar\\Image_::__SizeMedium__ - 70x70px
 - _EasyGravatar\\Image_::__SizeLarge__ - 120x120px
 
 
### Rating
 'Gravatar allows users to self-rate their images so that they can indicate if an image is appropriate for a certain audience.'. To change rating use `EasyGravatar\Image::setRating($rating);` method. By default rating is set to `g`.
 
	$img->setRating('pg');
	
Constants:

 - _EasyGravatar\\Image_::**RATING_G** - Suitable for display on all websites with any audience type.
 - _EasyGravatar\\Image_::**RATING_PG** - May contain rude gestures, provocatively dressed individuals, the lesser swear words, or mild violence.
 - _EasyGravatar\\Image_::**RATING_R** - May contain such things as harsh profanity, intense violence, nudity, or hard drug use.
 - _EasyGravatar\\Image_::**RATING_X** - May contain hardcore sexual imagery or extremely disturbing violence.


### Default
 When avatar you want fet doesn't exists, default Gravatar image is showed. You can change default image to yours, or one of Gravatar's builded in. To change it use `EasyGravatar\Image::setDefault($default);`.
 
	$img->default('http://example.com/avatars/default.png');
	
Constants:

 - _EasyGravatar\\Image_::__MysteryMan__ - A simple, cartoon-style silhouetted outline of a person (does not vary by email hash)
 - _EasyGravatar\\Image_::__Identicon__ - A geometric pattern based on an email hash
 - _EasyGravatar\\Image_::__Monster__ - A generated 'monster' with different colors, faces, etc
 - _EasyGravatar\\Image_::__Wavatar__ - Generated faces with differing features and backgrounds
 - _EasyGravatar\\Image_::__Retro__ - Awesome generated, 8-bit arcade-style pixelated faces
 
### Force default
 If you you want to make default image always loaded, forceDefault can be useful here. To use forceDefault use `EasyGravatar\Image::setForcedDefault($bool);`.
 
	$img->setForcedDefault(true);
	
### Secure
 If you want to display avatar from page that is served over SSL, this function can be useful. To use it, set secure to true by `EasyGravatar\Image::useSecure(true);`
 
	$img->useSecure(true);
	
- - -

 To make code cleaner and shorter EasyGravatar allows you to set parameters in chain. Here's some simple example:
 
	$img->setSize(50)
		->setForcedDefault(true)
		->setRating( EasyGravatar\Image::RATING_PG )
		->useSecure(true);



## Profile
 Easy gravatar provides simple support for profile requests too. (Yay! :D)

### Basics
 EasyGravatar user profile data gathering is based on getting element from response array with specifing path. Simplest way to get user's profile data you just need to:
 
 1. Include `path/to/EasyGravatar/Profile.php`
 2. Create `EasyGravatar\Profile` object
 3. Load user data using `EasyGravatar\Profile::loadProfile();`
 4. Get data that you are interested using `EasyGravatar\Profile::get($path);`
 
 
In PHP it's looking like that:

    # Step 1
     include 'path/to/EasyGravatar/Profile.php';
     	 
    # Step 2
     $profile = new Profile('usermail@example.com');
     	 
    # Step 3
     $profile->loadProfile();
     	 
    # Step 4
     echo $profile->getDisplayName();
	 

### Profile fields
 To get user's data you can use `EasyGravatar\Profile` getters. Here's list of them:

__Basic:__

 - ID - `EasyGravatar\Profile::getID();`      
 - Hash - `EasyGravatar\Profile::getHash();`
 - Requested Hash - `EasyGravatar\Profile::getRequestedHash();`
 - Profile URL - `EasyGravatar\Profile::getURL();`
 - Preferred Username - `EasyGravatar\Profile::getPreferredUsername();`
 - Thumbnail/Avatar URL - `EasyGravatar\Profile::getAvatar();`  
 - Display Name - `EasyGravatar\Profile::getDisplayName();`  
 
__User info:__

 - About - `EasyGravatar\Profile::getAboutInfo();`
 - Given name - `EasyGravatar\Profile::getGivenName();`
 - Family name - `EasyGravatar\Profile::getFamilyName();`
 - Formatted name - `EasyGravatar\Profile::getFormattedName();` 
 - Location - `EasyGravatar\Profile::getLocation();` 
 
__Background:__  
 - Color - `EasyGravatar\Profile::getBackgroundColor();` 
 - Position - `EasyGravatar\Profile::getBackgroundPosition();` 
 - Repeat - `EasyGravatar\Profile::getBackgroundRepeat();` 
 - URL - `EasyGravatar\Profile::getBackgroundURL();` 
 
__Array-Returns:__
 
 - Photos - `EasyGravatar\Profile::getPhotos();`
 - URLs - `EasyGravatar\Profile::getURLs();`
 - Accounts - `EasyGravatar\Profile::getAccounts();`
 - IMs - `EasyGravatar\Profile::getIMs();`
 - Mails - `EasyGravatar\Profile::getMails();`

 > __TIP:__
 > *If you don't want to use `EasyGravatar\Classname` all time, type `use EasyGravatar\classname;` on beggining of your script.*
 
  &nbsp;
 
 > __TIP:__
 > *If you want to print user name, use `EasyGravatar\Profile::getDisplayName()`.*
 
Array-return getters are returning array of items, for example `EasyGravatar\Profile::getPhotos()` will return __array__ of photos.



### Arrays build

#### Photos

Each photo array have some fields:


    {
	    value: [mixed],
	    [type: [mixed]]
    }
    

__value__ - URL to photo

__type__ - Contains "thumbnail" if photo is user's avatar *[oprional]*



#### Emails


	{
		value: [mixed],
		primary: [bool as string]
    }
    

__value__ - E-Mail adress

__primary__ - Is that email-primary?. Bool represented as string. Instead of true/false there is "true"/"false". So using `if` statement will always return **true**.

    
    
    
#### URLs

    {
		value: [mixed],
		title: [mixed]
    }
    

__value__ - URL

__title__ - URL name



#### IMs


    {
		type: [mixed]
		value: [mixed]
    }


__type__ - IM type

__value__ - User's IM id



#### Accounts

    {
		domain: [mixed],
		display: [mixed],
		url: [mixed]
		username\userid: [mixed/int],
		verified: [bool as string],
		shortname: [mixed]		
    }
    

__domain__ - Contains domain only

__display__ - Accound display, for example: for twitter it could be @name

__url__ - URL to you accound on the website

__username__/__userid__ - User name or id

__verified__ - Is user verified 

__shortname__ - provider short name


CHANGELOG.markdown
## 0.1.2 

 - __README updated__
 - Added new `Gravatar\Image::getImage()` return type: __Markdown__
 - Deleted all `Gravatar\Profile` constans, added getters
 - Added background getters
 - `Gravatar\Profile::loadProfile()` now throws `RuntimeException` when failed to load


lib/EasyGravatar/Profile.php
<?php
/**
 *      Gravatar Profile
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; either version 2 of the License, or
 *      (at your option) any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *      MA 02110-1301, USA.
 *
 *      @author Robik <szadows@gmail.com>
 *      @name Gravatar
 *      @license See LICENSE file included to package
 */

namespace EasyGravatar;

class Profile
{ 

    /**
     * Contains mail hash
     *
     * @var <mixed> Mail hash
     */
    protected $hash;

    /**
     * Contains user profile data
     *
     * @var <mixed[]> User profile data
     */
    protected $data;



    public function __construct($hashOrMail)
     {
        # If mail is already hashed
        if(strpos($hashOrMail, '@') === false)
            $this->hash = $hashOrMail;
        else
            $this->hash = md5(strtolower( trim($hashOrMail) ));
     }

     /**
      * Gets array element by path
      *
      * @param <mixed> Path
      *
      * @return <mixed[]>
      */
     public function get($path)
     {
         $elems = explode('/', $path);
         $temp = $this->data;

         foreach($elems as $elem)
         {
             if(!empty($elem))
                $temp = $temp[$elem];
         }

         return $temp;
     }

     /**
      * Loads user profile data
      *
      * @return <GEasyGravatar\Profile>
      */
     public function loadProfile()
     {
         $url = 'http://www.gravatar.com/' . $this->hash.'.php';
         
         $res = file_get_contents($url);

         if($res === false)
             throw new RuntimeException('Failed to load Profile.');

         $res = unserialize($res);
         $this->data = $res['entry'][0]; 

         return $this;
     }


     /**
      * Gets user's ID
      *
      * @return <int> User's ID
      */
     public function getID()
     {
         return $this->get('id');
     }

    /**
     * Gets user's e-mail hash
     *
     * @return <mixed> User e-mail hash
     */
    public function getHash()
    {
        return $this->get('hash');
    }

    /**
     * Gets requested hash
     *
     * @return <mixed> Requested hash
     */
    public function getRequestedHash()
    {
        return $this->get('requestedHash');
    }

    /**
     * Gets user's profile URL
     *
     * @return <mixed> User's profile URL
     */
    public function getProfileURL()
    {
        return $this->get('profileUrl');
    }

    /**
     * Gets user's preferred username
     *
     * @return <mixed> Preferred username
     */
    public function getPreferredUsername()
    {
        return $this->get('preferredUsername');
    }

    /**
     * Gets user's thumbail URL
     *
     * @return <mixed> User's avatar url
     */
    public function getAvatar()
    {
        return $this->get('thumbnailUrl');
    }

    /**
     * Gets user's photos
     *
     * @return <mixed[]> User photos
     */
    public function getPhotos()
    {
        return $this->get('photos');
    }

    /**
     * Gets user's display name
     *
     * @return <mixed> User's display name
     */
    public function getDisplayName()
    {
        return $this->get('displayName');
    }

    /**
     * Gets user's URLs
     *
     * @return <mixed[]> User URLs
     */
    public function getURLs()
    {
        return $this->get('urls');
    }


    /**
     * Gets user's about info
     *
     * @return <mixed> User's about info
     */
    public function getAboutInfo()
    {
        return $this->get('about');
    }

    /**
     * Gets user's given name
     *
     * @return <mixed> User's given name
     */
    public function getGivenName()
    {
        return $this->get('name\givenName');
    }

    /**
     * Gets user's family name
     *
     * @return <mixed> User's family name
     */
    public function getFamilyName()
    {
        return $this->get('name\familyName');
    }

    /**
     * Gets user's formatted name
     *
     * @return <mixed> User's formatted name
     */
    public function getFormattedName()
    {
        return $this->get('name\formatted');
    }

    /**
     * User's instant messagers accounts
     *
     * @return <mixed[]> User's instant messagers accounts
     */
    public function getIMs()
    {
        return $this->get('ims');
    }

    /**
     * Gets user's e-mails
     *
     * @return <mixed[]> User's emails
     */
    public function getMails()
    {
        return $this->get('emails');
    }

    /**
     * Gets user's current location
     *
     * @return <mixed> User's current location
     */
    public function getLocation()
    {
        return $this->get('currentLocation');
    }

    /**
     * Gets user's accounts
     *
     * @return <mixed[]> User's accounts
     */
    public function getAccounts()
    {
        return $this->get('accounts');
    }

    

    /**
     * Gets user's background color
     *
     * @return <mixed>
     */
    public function getBackgroundColor()
    {       
         return $this->get('profileBackground\color');
    }  
    
    /**
     * Gets user's background position
     *
     * @return <mixed>
     */
    public function getBackgroundPosition()
    {       
         return $this->get('profileBackground\position');
    }

    /**
     * Gets user's background repeat
     *
     * @return <mixed>
     */
    public function getBackgroundRepeat()
    {
         return $this->get('profileBackground\repeat');
    }

    /**
     * Gets user's background URL
     *
     * @return <mixed>
     */
    public function getBackgroundURL()
    {
         return $this->get('profileBackground\URL');
    }  

}

?>


lib/EasyGravatar/Image.php
<?php
/**
 *      Gravatar Image
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; either version 2 of the License, or
 *      (at your option) any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *      MA 02110-1301, USA.
 *
 *      @author Robik <szadows@gmail.com>
 *      @name Gravatar
 *      @license See LICENSE file included to package
 */

namespace EasyGravatar;

 class Image
 {
     /**
      * Suitable for display on all websites with any audience type.
      */
     const RATING_G = 'g';

     /**
      * May contain rude gestures, provocatively dressed individuals, the lesser swear words, or mild violence.
      */
     const RATING_PG = 'pg';

     /**
      * May contain such things as harsh profanity, intense violence, nudity, or hard drug use.
      */
     const RATING_R = 'r';

     /**
      * May contain hardcore sexual imagery or extremely disturbing violence.
      */
     const RATING_X = 'x';


     /**
      * A simple, cartoon-style silhouetted outline of a person (does not vary by email hash)
      */
     const MysteryMan = 'mm';

     /**
      * A geometric pattern based on an email hash
      */
     const Identicon = 'identicon';

     /**
      * A generated 'monster' with different colors, faces, etc
      */
     const Monster = 'monsterid';

     /**
      * Generated faces with differing features and backgrounds
      */
     const Wavatar = 'wavatar';

     /**
      * Awesome generated, 8-bit arcade-style pixelated faces
      */
     const Retro = 'retro';



     /**
      * Returns image URL
      */
     const Plain = 'plain';

     /**
      * Returns image as HTML's IMG tag
      */
     const HTML = 'html';

     /**
      * Returns image as BBCode IMG tag
      */
     const BB = 'bb';

     /**
      * Returns image in Markdown syntax
      */
     const Markdown = 'markdown';


     /**
      * Defines small avatar size
      */
     const SizeSmall = 30;

     /**
      * Defines medium avatar size
      */
     const SizeMedium = 70;

     /**
      * Defines large avatar size
      */
     const SizeLarge = 120;



     /**
      * Contains mail
      *
      * @var <mixed>
      */
     protected $hash;

     /**
      * Image size
      *
      * @var <int>
      */
     protected $size = 80;

     /**
      * Contains default image (occurs when image don't exists)
      *
      * @var <mixed> Path/Build-in
      */
     protected $default = 'identicon';

     /**
      * Image generated even with not exists?
      *
      * @var <bool>
      */
     protected $forceDefault = false;

     /**
      * Image rating
      *
      * @var <mixed>
      */
     protected $rating = 'g';

     /**
      * Contains secure use
      *
      * @var <bool> Secure use
      */
     protected $secure;



     public function __construct($hashOrMail)
     {
        # If mail is already hashed
        if(strpos($hashOrMail, '@') === false)
            $this->hash = $hashOrMail;
        else
            $this->hash = md5(strtolower( trim($hashOrMail) ));
     }


     /**
      * Sets gravatar size
      *
      * @param <int> Size
      *
      * @return <EasyGravatar\Image>
      */
     public function setSize($size)
     {
         $this->size = (int)$size;

         return $this;
     }


     /**
      * Sets max rating
      *
      * @param <mixed> Rating
      *
      * @return <EasyGravatar\Image>
      */
     public function setRating($rating)
     {
         $this->rating = $rating;

         return $this;
     }


     /**
      * Sets default image
      *
      * @param <mixed> Path to default image/default
      *
      * @return <EasyGravatar\Image>
      */
     public function setDefault($default)
     {
         $this->default = urlencode($default);
         return $this;
     }

     
     /**
      * Sets forcedDefautt use
      *
      * @param <bool> Use
      *
      * @return <EasyGravatar\Image>
      */
     public function setForcedDefault($bool)
     {
         $this->forcedDefault = ($bool ? 'y' : '');
         return $this;
     }


     /**
      * Use secure use
      *
      * @param <bool> Use
      *
      * @return <EasyGravatar\Image>
      */
     public function useSecure($bool)
     {
         $this->secure = $bool;
         return $this;
     }


     /**
      * Gets user image
      *
      * @param <mixed> Return Content type
      * @param <mixed> Image alt (used only in HTML content)
      *
      * @return <mixed>
      */
     public function getImage($type = 'plain', $alt = 'Avatar')
     {
         if($this->secure != NULL)
            $url = 'https://secure.gravatar.com/avatar/'.$this->hash;
         else
            $url = 'http://www.gravatar.com/avatar/'.$this->hash;

         $params = '?size=' . $this->size;

         if($this->default != NULL)
            $params .= '&default=' . $this->default;
         
         $params .= '&rating=' . $this->rating;

         if($this->forceDefault != NULL)
            $params .= '&forcedefault=y';

         if($type == 'plain')
             return $url.$params;
         if($type == 'html')
             return '<img src="' . $url.$params . '" alt="'.$alt.'" />';
         if($type == 'bb')
             return '[img]' . $url.$params . '[/img]';
         if($type == 'markdown')
             return '![' . $alt . '](' . $url.$params . ')';

     }
 }

?>

